### ISSUE TITLE
Undefined symbols "Moose::perf_log" for architecture arm64 linking libmoose

### ISSUE THREAD

**Issue by Jacobfaib** [idaholab/moose#20276]
Error Message:
```
Undefined symbols for architecture arm64:
  "Moose::perf_log", referenced from:
      CheckOutputAction::checkPerfLogOutput() in actions_Unity.arm-apple-darwin21.3.0.opt.o
      Console::~Console() in outputs_Unity.arm-apple-darwin21.3.0.opt.o
      Console::output(MooseEnumItem const&) in outputs_Unity.arm-apple-darwin21.3.0.opt.o
      MooseApp::setupOptions() in MooseApp.arm-apple-darwin21.3.0.opt.o
ld: symbol(s) not found for architecture arm64
```
Full error message [link_err.txt](https://github.com/idaholab/moose/files/8043313/link_err.txt)
As well as attempted command [link_out.txt](https://github.com/idaholab/moose/files/8043349/link_out.txt)

This is on latest `master`:
```
$ git branch
* master
$ git log
commit 26e85a13f385a14bbc39508297586d03cc095f36 (HEAD -> master, origin/master)
Merge: b4cd6c4323 a0de0c877d
Author: moosetest <bounces@inl.gov>
Date:   Wed Feb 9 16:22:33 2022 -0700

    Merge commit 'a0de0c877d2647785966c81bdfaaf03209e4e886'
<...>
```

System Information:
```
$ uname -a
Darwin MacBook-Pro.local 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000 arm64

$ which mpicc
/Users/jacobfaibussowitsch/soft/bin/mpicc
$ mpicc -show
gcc -fPIC -fstack-protector -fno-stack-check -Qunused-arguments -g3 -O0 -fno-common -Wno-implicit-function-declaration -Wl,-flat_namespace -Wl,-commons,use_dylibs -I/Users/jacobfaibussowitsch/soft/include -L/Users/jacobfaibussowitsch/soft/lib -lmpi -lpmpi
$ mpicc --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 13.0.0 (clang-1300.0.29.30)
Target: arm64-apple-darwin21.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

$ which mpicxx
/Users/jacobfaibussowitsch/soft/bin/mpicxx
$ mpicxx -show
g++ -fstack-protector -fno-stack-check -g -O0 -fPIC -std=gnu++17 -Wl,-flat_namespace -Wl,-commons,use_dylibs -I/Users/jacobfaibussowitsch/soft/include -L/Users/jacobfaibussowitsch/soft/lib -lmpicxx -lmpi -lpmpi
$ mpicxx --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 13.0.0 (clang-1300.0.29.30)
Target: arm64-apple-darwin21.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

$ which mpif90
/Users/jacobfaibussowitsch/soft/bin/mpif90
$ mpif90 -show
gfortran -fPIC -ffree-line-length-0 -g -O0 -fallow-argument-mismatch -Wl,-flat_namespace -Wl,-commons,use_dylibs -I/Users/jacobfaibussowitsch/soft/include -I/Users/jacobfaibussowitsch/soft/include -L/Users/jacobfaibussowitsch/soft/lib -lmpifort -lmpi -lpmpi
$ mpif90 --version
GNU Fortran (Homebrew GCC 11.2.0_3) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Misc Info:

- I had very similar problems linking `libgtest` and `libhit`, which I solved by changing the compiler and link target to C++ rather than C (maybe I mis-configured something, but should this not be the default anyways as `libgtest` and `libhit` are C++ libraries themselves). For `libhit` I also had to include `$(hit_LIB)` in the `$(moose_LIB)` target as it was missing.
```
diff --git a/framework/moose.mk b/framework/moose.mk
index 72af8a2550..3f638f80d3 100644
--- a/framework/moose.mk
+++ b/framework/moose.mk
@@ -315,20 +315,20 @@ $(pcre_LIB): $(pcre_objects)

 $(gtest_LIB): $(gtest_objects)
        @echo "Linking Library "$@"..."
-       @$(libmesh_LIBTOOL) --tag=CC $(LIBTOOLFLAGS) --mode=link --quiet \
-         $(libmesh_CC) -o $@ $(gtest_objects) $(EXTERNAL_FLAGS) -rpath $(gtest_DIR)
+       @$(libmesh_LIBTOOL) --tag=CXX $(LIBTOOLFLAGS) --mode=link --quiet \
+         $(libmesh_CXX) -o $@ $(gtest_objects) $(EXTERNAL_FLAGS) -rpath $(gtest_DIR)
        @$(libmesh_LIBTOOL) --mode=install --quiet install -c $(gtest_LIB) $(gtest_DIR)

 $(hit_LIB): $(hit_objects)
        @echo "Linking Library "$@"..."
-       @$(libmesh_LIBTOOL) --tag=CC $(LIBTOOLFLAGS) --mode=link --quiet \
+       @$(libmesh_LIBTOOL) --tag=CXX $(LIBTOOLFLAGS) --mode=link --quiet \
          $(libmesh_CXX) $(CXXFLAGS) $(libmesh_CXXFLAGS) -o $@ $(hit_objects) $(libmesh_LDFLAGS) $(libmesh_LIBS) $(EXTERNAL_FLAGS) -rpath $(HIT_DIR)
        @$(libmesh_LIBTOOL) --mode=install --quiet install -c $(hit_LIB) $(HIT_DIR)

 $(moose_LIB): $(moose_objects) $(pcre_LIB) $(gtest_LIB) $(hit_LIB) $(pyhit_LIB)
        @echo "Linking Library "$@"..."
-       @$(libmesh_LIBTOOL) --tag=CXX $(LIBTOOLFLAGS) --mode=link --quiet \
-         $(libmesh_CXX) $(CXXFLAGS) $(libmesh_CXXFLAGS) -o $@ $(moose_objects) $(pcre_LIB) $(png_LIB) $(libmesh_LDFLAGS) $(libmesh_LIBS) $(EXTERNAL_FLAGS) -rpath $(FRAMEWORK_DIR)
+       @$(libmesh_LIBTOOL) --tag=CXX $(LIBTOOLFLAGS) --mode=link \
+         $(libmesh_CXX) $(CXXFLAGS) $(libmesh_CXXFLAGS) -o $@ $(moose_objects) $(pcre_LIB) $(png_LIB) $(hit_LIB) $(libmesh_LDFLAGS) $(libmesh_LIBS) $(EXTERNAL_FLAGS) -rpath $(FRAMEWORK_DIR)
        @$(libmesh_LIBTOOL) --mode=install --quiet install -c $(moose_LIB) $(FRAMEWORK_DIR)

 ifeq ($(MOOSE_HEADER_SYMLINKS),true)
```
But I noted that the `$(moose_LIB)` target was already using the C++ compiler to compiler and link, so not sure what is different here.

- LibMesh was install via MOOSE with the following configure arguments:
```
#!/usr/bin/env bash
export METHODS="devel"
export PETSC_ARCH=arch-moose
export MOOSE_JOBS=${MOOSE_JOBS:-6}
./update_and_rebuild_libmesh.sh --with-thread-model=auto \
                                --disable-timestamps \
                                --enable-perflog \
                                --with-gdb-command=lldb \
                                --with-cc=mpicc \
                                --with-cxx=mpicxx \
                                --with-fc=mpif90 \
                                --with-f77=mpif77
```
- PETSc came from a local install

**Comment by Jacobfaib** [comment]
After trying multiple things the only thing I could get to work was to simply comment out the offending lines in
```
framework/src/actions/CheckOutputAction.C
framework/src/base/MooseApp.C
framework/src/outputs/Console.C
```
After that, `libmoose` did successfully link, but now `libmoose_test.la` choked, complaining that pretty much all symbols in `libmoose` were undefined... Fix for this was yet another makefile change. Specifically adding `$(moose_LIB)` as a link dependency to the app, and adding the app itself as a link dependency to the test variant.
```
diff --git a/framework/app.mk b/framework/app.mk
index 77120a4fd9..11f10416af 100644
--- a/framework/app.mk
+++ b/framework/app.mk
@@ -334,7 +334,7 @@ $(app_LIB): curr_libs := $(depend_libs_flags)
 $(app_LIB): $(app_HEADER) $(app_plugin_deps) $(depend_libs) $(app_objects) $(ADDITIONAL_DEPEND_LIBS)
        @echo "Linking Library "$@"..."
        @$(libmesh_LIBTOOL) --tag=CXX $(LIBTOOLFLAGS) --mode=link --quiet \
-         $(libmesh_CXX) $(CXXFLAGS) $(libmesh_CXXFLAGS) -o $@ $(curr_objs) $(libmesh_LDFLAGS) $(EXTERNAL_FLAGS) -rpath $(curr_dir)/lib $(curr_libs)
+         $(libmesh_CXX) $(CXXFLAGS) $(libmesh_CXXFLAGS) -o $@ $(curr_objs) $(libmesh_LDFLAGS) $(EXTERNAL_FLAGS) $(moose_LIBS) -rpath $(curr_dir)/lib $(curr_libs)
        @$(libmesh_LIBTOOL) --mode=install --quiet install -c $@ $(curr_dir)/lib

 ifeq ($(BUILD_TEST_OBJECTS_LIB),no)
@@ -351,7 +351,7 @@ $(app_test_LIB): curr_libs := $(depend_libs_flags)
 $(app_test_LIB): $(app_HEADER) $(app_plugin_deps) $(depend_libs) $(app_test_objects) $(ADDITIONAL_DEPEND_LIBS)
        @echo "Linking Library "$@"..."
        @$(libmesh_LIBTOOL) --tag=CXX $(LIBTOOLFLAGS) --mode=link --quiet \
-         $(libmesh_CXX) $(CXXFLAGS) $(libmesh_CXXFLAGS) -o $@ $(curr_objs) $(libmesh_LDFLAGS) $(EXTERNAL_FLAGS) -rpath $(curr_dir)/lib $(curr_libs)
+         $(libmesh_CXX) $(CXXFLAGS) $(libmesh_CXXFLAGS) -o $@ $(curr_objs) $(app_LIB) $(libmesh_LDFLAGS) $(EXTERNAL_FLAGS) -rpath $(curr_dir)/lib $(curr_libs)
        @$(libmesh_LIBTOOL) --mode=install --quiet install -c $@ $(curr_dir)/lib
 endif
```
I am starting to think that perhaps I have misconfigured something, as there should be no way that either of these make targets should work for *anyone* in their current state.

**Comment by cticenhour** [comment]
First of all, we don't currently officially support ARM64 quite yet. We also don't support homebrew stacks officially (though we do try to help where we can here). However, our pre-release conda package is performing very well so far, and it successfully passes our MOOSE framework tests (and almost all module tests as of now - there are a few stragglers). We don't see these symbol errors using that stack. So conda (installing just `moose-mpich` and `moose-tools` since you are building your own PETSc and libMesh) could be an option for you to get moving with a compatible stack. 

I see that you are using homebrew packages to provide compiler wrappers and clang 13.0.0. We have tested MOOSE on M1 using clang 12.0.1 and MPICH 3.4.2 (again using conda, not homebrew), so you might try your build again using that compiler version. Another related source of error here could be the homebrew configuration for your MPI compiler wrapper....you might take a look at how we build MPICH (see the build script for our conda package [here](https://github.com/idaholab/moose/blob/next/conda/mpich/build.sh) for configure info) and cross-reference that to the homebrew script (if you are using MPICH). 

I am also puzzled by your issues with `libgtest` and `libhit`. Again, we haven't seen that quite yet. It does make me think something is mis-configured higher up in your stack. I might attempt a local upgrade of our packages to clang 13.0.0 to see if I can replicate. I will report back here if I do. 

Thank you for reporting this, even if we don't currently support ARM64 or homebrew. I want to tag @dschwen and @milljm, as they have both been working alongside me in our ARM64 support effort. They might also have some ideas to share beyond what I've put up here.

**Comment by cticenhour** [comment]
Well, I just noticed that you aren't using homebrew for all of this, only gfortran. So might have to disregard some of what I'm saying here (at least in terms of the compiler itself). Are you building your own MPI wrapper?

**Comment by Jacobfaib** [comment]
Note with all of the above diffs, I was able to get `libmoose` and friends, as well as an external app to build and link together. The one issue that persists is that I still don't know how to solve the original undefined `Moose::perf_log` problem without commenting it out. Spelunking through the headers it seems to be defined in `framework/include/base/Moose.h:134`
```
/**
 * Perflog to be used by applications.
 * If the application prints this in the end they will get performance info.
 */
extern PerfLog perf_log;
```
where `PerfLog` (which is in fact secretly `libMesh::PerfLog`) is defined in `include/utils/perf_log.h:126`. Trying to find the actual definition of `perf_log` then under `framework/` yields
```
moose/framework$ rg perf_log
src/base/MooseApp.C
713:    //Moose::perf_log.disable_logging();
720:    //Moose::perf_log.disable_logging();
727:    //Moose::perf_log.disable_logging();
844:    //Moose::perf_log.disable_logging();
856:    //Moose::perf_log.disable_logging();
877:    //Moose::perf_log.disable_logging();
898:    //Moose::perf_log.disable_logging();
911:    //Moose::perf_log.disable_logging();
989:    //Moose::perf_log.disable_logging();

src/actions/CheckOutputAction.C
146:    //Moose::perf_log.disable_logging();

src/actions/CommonOutputAction.C
116:      "print_perf_log", false, "Use perf_graph instead!", "Use perf_graph instead!");
229:        (getParam<bool>("perf_graph") || getParam<bool>("print_perf_log") ||

src/outputs/Console.C
73:  params.addDeprecatedParam<bool>("perf_log",
79:      "perf_log_interval",
86:      "Print the libMesh performance log header (requires that 'perf_log = true')",
131:  params.addParamNamesToGroup("perf_log solve_log perf_header", "Perf Log");
172:    _perf_log(getParam<bool>("perf_log")),
173:    _perf_log_interval(getParam<unsigned int>("perf_log_interval")),
174:    _solve_log(isParamValid("solve_log") ? getParam<bool>("solve_log") : _perf_log),
176:    _perf_header(isParamValid("perf_header") ? getParam<bool>("perf_header") : _perf_log),
204:  if (!_pars.isParamSetByUser("perf_log") && common && common->getParam<bool>("print_perf_log"))
206:    _perf_log = true;
234:  //Moose::perf_log.disable_logging();
241:  // Only allow the main app to change the perf_log settings.
353:    if (_perf_log_interval && _t_step % _perf_log_interval == 0)
354:      //write(Moose::perf_log.get_perf_info(), false);

doc/content/syntax/Outputs/index.md
81:    perf_log = true
117:  print_perf_log = true       # enable the performance log

include/base/Moose.h
12:#include "libmesh/perf_log.h"
134:extern PerfLog perf_log;

include/outputs/Console.h
181:  bool _perf_log;
184:  unsigned int _perf_log_interval;
```
I don't see a definition of it. The only possible definition I do see is under `test/`:
```
moose/test$ rg perf_log
src/main.C
18:PerfLog Moose::perf_log("Moose Test");   // HERE

tests/postprocessors/print_perf_data/tests
17:    # Make sure that the perf_log table doesn't just report
```
but isn't this be built *after* `libmoose` as part of `libmoose_test`?

**Comment by cticenhour** [comment]
I don't work with our framework/app Makefiles, so I admit I'm flying a bit blind here - bear with me. 

> otherwise how would
> ```
> $(libmesh_CXX) $(CXXFLAGS) $(libmesh_CXXFLAGS) -o $@ \
> $(moose_objects) $(pcre_LIB) $(png_LIB) \
> $(libmesh_LDFLAGS) $(libmesh_LIBS) $(EXTERNAL_FLAGS) -rpath $(FRAMEWORK_DIR)
> ```
> find the recently built libhit? Unless it is supposed to be bundled into $(libmesh_LIBS)?

Looking at what you've wrote as well as what I see in my build of MOOSE, `libgtest` and `libhit` both exist within `FRAMEWORK_DIR/contrib/gtest` and `FRAMEWORK_DIR/contrib/hit` after being built, and we refer to `FRAMEWORK_DIR` in the targets in the `makefile` (as in your quote above). So I don't understand why they aren't being found here in your case. 

Our `hit` target  in `moose.mk` does include `-std=c++11` and `-lstdc++`, but it doesn't appear to be involved in the target you are referencing. 

> `clang++` will automatically include `-lstdc++` etc and linker paths, `clang` will not.

On my Mac, `CXXFLAGS` is set and provides `-stdlib=libc++` and `-std=c++14`, which would apply in those targets. This is set via conda alongside the activation of our compiler/MPICH package. I am unsure how these are provided in the [fully manual install](https://mooseframework.inl.gov/getting_started/installation/manual_installation_llvm.html) though off the top of my head, as we don't set that environment variable. 

For your perf_log question I am deferring to @friedmud, who has done a lot of work related to the performance log system in MOOSE and may shed some light.

**Comment by Jacobfaib** [comment]
> Looking at what you've wrote as well as what I see in my build of MOOSE, libgtest and libhit both exist within FRAMEWORK_DIR/contrib/gtest and FRAMEWORK_DIR/contrib/hit after being built, and we refer to FRAMEWORK_DIR in the targets in the makefile (as in your quote above). So I don't understand why they aren't being found here in your case.

It's not that it isn't *found* per-se (both of those directories are as you mention included in the link invocation), it's that `libhit` is not requested during the link. That is, when I compile the link command is missing
```
/path/to/libhit.[dylib|so]
// or
-lhit
```
Adding `$(hit_LIB)` adds the former to the link command. Per your next point about storing stuff in `CXXFLAGS` perhaps you have this in `LDFLAGS`, `LIBRARY_PATH` or `LD_LIBRARY_PATH`?

> On my Mac, CXXFLAGS is set and provides -stdlib=libc++ and -std=c++14, which would apply in those targets. This is set via conda alongside the activation of our compiler/MPICH package. I am unsure how these are provided in the [fully manual install](https://mooseframework.inl.gov/getting_started/installation/manual_installation_llvm.html) though off the top of my head, as we don't set that environment variable.

Yeah `CXXFLAGS` (and for that matter `CFLAGS`, `CXXOPTFLAGS`, and `COPTFLAGS`) is empty for me, both normally and during the build.

**Comment by friedmud** [comment]
The missing `perf_log` symbol would happen because you're not linking a `main.o`.  The PerfGraph is defined in the _one_ `main.C` that you are linking for this application.  Look at `Mastodon` `main.C` here: https://github.com/idaholab/mastodon/blob/devel/src/main.C#L8

The only way `Moose::perf_log()` is undefined is if you deleted it from your `main.C` or you forgot to link a `main.C` (which is not possible with our default makefiles that you get with `stork`).

**Comment by Jacobfaib** [comment]
But I'm not linking an app in the above. I am trying to build `libmoose` as a prerequisite to building `libmy_app`. For reference I am running
```
moose/test $ make -j np
```
Which builds `libmoose` then `libmoose_test`. And the former (which is the one that complains about a missing `perf_log`) does not link in a `main.o` because that isn't added until `libmoose_test`.

**Comment by Jacobfaib** [comment]
Because it is ODR-used in the code (by calling member functions), and hence the linker would need to find it somewhere in order to resolve the symbol.

**Comment by friedmud** [comment]
But it's been this way for ~12 years.  If there were an issue with having that symbol in `main.C` it would have shown up before now.

**Comment by friedmud** [comment]
We need to figure out why your particular combination of OS/compilers is baulking at this when none ever has before.

**Comment by Jacobfaib** [comment]
> Would you mind using our normal install instructions and our package to see if you still have the issue?

You mean the conda install? For reference to get the above I was following https://mooseframework.inl.gov/getting_started/installation/offline_installation.html

**Comment by Jacobfaib** [comment]
>  Even on ARM.

FWIW I tried all the above on my old Intel mac about a week ago with the same results. It had an identical setup (building moose from source, with identical PETSc install also providing mpi wrappers). I have since sent that one back though so I cannot try and reproduce to get exact measurements :(. 

So perhaps more a Apple clang/lld issue?

**Comment by cticenhour** [comment]
> So perhaps more a Apple clang/lld issue?

It could be, but I am unsure how to address an issue there.

> For reference to get the above I was following https://mooseframework.inl.gov/getting_started/installation/offline_installation.html

To my knowledge, the Offline Installation instructions is intended for HPC systems, generally. So there could be a configuration step or suggestion that works perfectly well in a Linux environment but doesn't work here. If you want to keep building your own stack, you might try the ["Manual Installation LLVM"](https://mooseframework.inl.gov/getting_started/installation/manual_installation_llvm.html) instructions instead. 

As @friedmud suggested, I would definitely be interested first in seeing how your machine handles [the conda packages.](https://mooseframework.inl.gov/getting_started/installation/conda.html) If you still have an issue there, or see some build behavior you don't expect, then the local environment might be to blame. We pre-build petsc and libmesh for those, but if you just want the compiler stack and tools for running the test harness so that you can build your own, you just need to perform `conda install moose-mpich moose-tools` during that step in the instructions.

**Comment by Jacobfaib** [comment]
Sorry for long hiatus. I have since tried installing moose via the conda install path:
```
$ curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-MacOSX-arm64.sh
$ bash ./Mambaforge-MacOSX-arm64.sh -b -p `pwd` -u
$ export PATH=`pwd`/bin:$PATH
$ conda config --add channels https://conda.software.inl.gov/public
$ mamba init
$ mamba create --name moose -q -y
$ mamba activate moose
$ mamba install moose-tools moose-libmesh
$ mamba deactivate
$ mamba activate moose
```
then recompiled moose
```
moose/test $ unset MOOSE_DIR
moose/test $ make -j <np>
```
I then get hundreds of instances of the following warning:
```
ld: warning: direct access in function 'ArrayKernel::computeResidual()' from file '/Users/jacobfaibussowitsch/soft/moose/framework/build/unity_src/.libs/kernels_Unity.aarch64-apple-darwin20.0.0.devel.o' to global weak symbol 'typeinfo for MooseVariableFE<Eigen::Matrix<double, -1, 1, 0, -1, 1> >' from file '/Users/jacobfaibussowitsch/soft/moose/framework/build/unity_src/.libs/variables_Unity.aarch64-apple-darwin20.0.0.devel.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```
but moose ultimately compiles and links. Going over to my app then
```
soft/app_dir $ make clean
soft/app_dir $ make -j <np>
```
errors with
```
arm64-apple-darwin20.0.0-gfortran: warning: could not understand version 12.2
ld: -rpath can only be used when targeting Mac OS X 10.5 or later
collect2: error: ld returned 1 exit status
gmake: *** [/Users/jacobfaibussowitsch/soft/moose/framework/build.mk:319: /Users/jacobfaibussowitsch/soft/moose/modules/tensor_mechanics/test/plugins/elastic_incremental-devel.plugin] Error 1
```

**Comment by cticenhour** [comment]
> I then get hundreds of instances of the following warning:
> ```
> ld: warning: direct access in function...
> ```

This is a known warning, due to the fact that PETSc on M1 does not seem to support shared libraries. We are working through eliminating or minimizing these warnings by the time we get to full M1 support. 

Your other warning/error is something I have not seen before when building on M1 using our stack. Before I perform a local experiment here, could you do a clean of MOOSE as well and try rebuilding? If you don't have any un-committed changes in MOOSE, a `git clean -xfd` should do it quite nicely.

**Comment by Jacobfaib** [comment]
> due to the fact that PETSc on M1 does not seem to support shared libraries.

[This would be news to me](https://gitlab.com/petsc/petsc/-/jobs/2124624784). I think this is more likely caused by petscs visibility flags (which appear to be automatically added to the global list of build flags) interfering with other libraries. Specifically, it compiles all sources with `-fvisibility=hidden` by default.

> Before I perform a local experiment here, could you do a clean of MOOSE as well and try rebuilding?

Same thing.

Note since `gfortran-11` I've had to put
```
export MACOSX_DEPLOYMENT_TARGET=12.2
```
as an env variable. Compiling with the brew-supplied gfortran produces no errors.

**Comment by cticenhour** [comment]
> [Hmm, this would be news to me](https://gitlab.com/petsc/petsc/-/jobs/2124624784). 

Very interesting - I was concerned this issue was more a problem with our current compiler configuration than anything PETSc, and this confirms that. Thank you! 

> I think this is more likely caused by petscs visibility flags (which appear to be automatically added to the global list of build flags) interfering with other libraries. Specifically, it compiles all sources with `-fvisibility=hidden` by default.

I did not realize that was being automatically added. We modified some visibility settings for M1, but that would explain this.

> Note since `gfortran-11` I've had to put
> 
> ```
> export MACOSX_DEPLOYMENT_TARGET=12.2
> ```
> 
> as an env variable. Compiling with the brew-supplied gfortran produces no errors.

We've never had to provide that for the conda-forge gfortran provided in our package. 

At this point I will need to perform a local test using a clean application. What modules do you have turned on? The fact that you are seeing this with a `tensor_mechanics` plugin is baffling, as we've not seen that in our own CI or local builds (the latter as of Friday, in fact).

**Comment by Jacobfaib** [comment]
> What modules do you have turned on?

Only `tensor_mechanics`. I attached the [moose-generated makefile](https://github.com/idaholab/moose/files/8120521/Makefile.txt) for the app (the app itself has not been modified since it was generated by stork).

**Comment by GiudGiud** [comment]
I think this can be closed as resolved by our current M1 support ?

**Comment by Jacobfaib** [comment]
Last I checked (circa April) this issue was still not fixed -- and required the fixes/workarounds I suggested above -- but I can pull latest `master` and check again in the next few days.

Note PETSc has recently moved to [remove visibility rules from compiler flags](https://gitlab.com/petsc/petsc/-/merge_requests/5810), but AFAIK this only applies when compiling Kokkos. Stock PETSc compile flags will still include the visibility flags.
